### ISSUE TITLE
PLAN : TBD

### ISSUE THREAD

**Issue by powerswing** [ztsv-av/general_model#12]
# TODO

Go through each task in task list one by one

## PREPROC AND PERMUTATIONS

- [x] prepare a logical folder structure (like the one below or any similar which is easy to work with)
- train
  - image
    - dicom
    - jpg
      - resized
    - numpy
  - study
    - ...

- [x] convert data to .jpg and save data to **jpg** folder
- [x] since data does not have the same ratio (and we need squares), convert all images to the same 1:1 ratio
  - if the height is bigger than width, cut the residual of the height so it becomes the same as width. Do the same with the width if it's bigger. The algorithm may look like so
    - suppose height is 3050 and width is 2850. The difference is 3050-2850 = 200. Hence cut the top by 100 (200/2) and cut the bottom by 100 (200/20)
- [x] resize .jpg to 512x512 (maybe?) and save resized .jpg to **resized** folder and arrays to **numpy** folder

  > resize function must be implemented in **preprocessing.py**

- [x] create a function **normalizeImage()** that uses mean and standard deviation. This function is called in batch generator after all permutation functions. Permutation functions are applied with some probability but normalization is applied with probability one

- [x] implement random permutations in **permutations.py**
  - rotation by a random degree
  - blur
  - contrast
  - invert colors
  - some more from albumentations

  > since permutations are already coded in albumentations, make wrappers for them
  > each permutation must be tested with resized .jpg images from **jpg** folder

## FIT

 - [x] create generator instance with new permutations
 - [x] add a normalization function into generator so it is applied to each image in each batch

## POST-FIT

 - [x] change time callback format to "2021-07-12 17:30"
 - [x] preprocess test data
 - [x] understand if additional data can be used

> Train a model for one epoch on data of input shape 256x256x3, store the weights, load them back again, and try to train the loaded model on data of input shape 512x512x3

 - [x] create a dictionary with 'modelname' and 'maximum batch size' pairs
 - [x] fit all pretrained models on the **yes/no external data** you found and select the weights for the main train
 - [x] fit all models with bce using pretrained weights (where possible) on external data with bce loss
 - [x] add missing studies for test (40 studies are missing)
 - [x] preprocess test folder for prediction step
- [x] obtain predictions for all models using preprocessed test data and save them as **.npy**
- [x] compile submission files of three types
  - one containing 4 labels with respective confidence levels 
  - one containing labels where respective confidence levels are above 0.5
  - one containing only one label with the highest probability
- [x] upload all submissions and check kaggle score to select the best compile technique

Now, using this technique

- [x] select N-models for ensemble
- [x] average saved predictions of the selected models
- [x] compile submission files
- [x] upload these submissions and check kaggle score
- [ ] retrain 256x256 models and 512x512 models
- [ ] check their scores
- [ ] ensemble best models (the more the better as far as I understand (compare like 8 models ensemble and 4 models ensemble)

## BOX MODELS
 - [x] implement buildModel function (where you download model architecture and initialize pretrained weights)
 - [x] implement createTrainData function, which takes numpy path and returns image (1, width, height, 3), boxes (N_i, 4), class (N_i, num_classes), where N_i - num of boxes in an image. Remember - all in tensors! Then pass this data to train_step function
 - [x] implement train_step function, with all preprocessing, prediction, loss and gradients stuff. (check if preprocessing is somehow connected to the model you are using, if not - implement your own preprocessing function)
 - [ ] implement metrics to see how you are doing
 - [x] save the model like here: [Exporting a Trained Model](https://tensorflow-object-detection-api-tutorial.readthedocs.io/en/latest/training.html#exporting-a-trained-model)
 - [x] try to load the model like you did in buildModel function and do predictions. if works, you are a fucking genius
 - [x] fine-tune parameters.
 - [ ] train more object detection models (check submission score before doing that??)
 - [ ] check score of each model
 - [ ] do an ensemble of the best of them?
 

## FINAL IMPLEMENTATIONS
- [ ] predict using best classification and object detection ensembles
- [ ] check score on kaggle
- [ ] code submission algorithm in kaggle notebook (for hidden test)

**Comment by ztsv-av** [comment]
I think there is no need to save files as jpg. Just takes up disk space. At any time I can read numpy file and show it to you, if necessary. Checked numpy files myself, they look as intended.

**Comment by ztsv-av** [comment]
resized numpies to 512x512 and changed all necessary data in csv's for each file (box sizes and opacities). took awhile... I'll show you the code later. 
I'll implement resize function in preprocessing.py tomorrow.

**Comment by powerswing** [comment]
> I think there is no need to save files as jpg. Just takes up disk space. At any time I can read numpy file and show it to you, if necessary. Checked numpy files myself, they look as intended.

We need jpg files so we can see whether resize to square works as intended. Also whether box resize works as intended. jpg files don't weigh much anyway.

**Comment by ztsv-av** [comment]
Sh*t, test data that we have is the data that we can predict on to create correct code for kaggle notebook. It doesn't have any correct labels, only DICOMs, so it is basically useless.

**Comment by ztsv-av** [comment]
found 16000 external data images for training, where each label is covid-19 positive or negative. will show you.
